### PR TITLE
feat: [AXIMST-375] Course unit - Added functionality for copying and pasting xblocks and units

### PR DESCRIPTION
**Pull request by PKulkoRaccoonGang**
**YouTrack**
- https://youtrack.raccoongang.com/issue/AXIMST-356
- https://youtrack.raccoongang.com/issue/AXIMST-350
- https://youtrack.raccoongang.com/issue/AXIMST-428
- https://youtrack.raccoongang.com/issue/AXIMST-375

### CODE REVIEW

File: src/course-unit/__mocks__/clipboardResponse.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  locator: 'block-v1:edX+DemoX+Demo_Course+type@vertical+block@vertical_0270f6de40fc',
+  courseKey: 'course-v1:edX+L153+3T2023',
+  static_file_notices: {
```
Comment (ruzniaievdm) [inline src/course-unit/__mocks__/clipboardResponse.js:4]: [question] Do we need to use here snake_case?

File: src/course-unit/clipboard/paste-component/components/PasteComponentButton.jsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import { Button } from '@edx/paragon';
+import { ContentCopy as ContentCopyIcon } from '@edx/paragon/icons';
+
+import messages from '../messages';
+
+const PasteComponentButton = ({ handleCreateNewCourseXBlock }) => {
+  const intl = useIntl();
+  const { blockId } = useParams();
+
+  const handlePastXBlockComponent = () => {
```
Comment (ruzniaievdm) [inline src/course-unit/clipboard/paste-component/components/PasteComponentButton.jsx:13]:
```suggestion
  const handlePasteXBlockComponent = () => {
```

File: src/course-unit/clipboard/paste-component/components/PopoverContent.jsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+          <strong>{content.displayName}</strong>
+          {sourceEditUrl && (
+            <Icon className="clipboard-popover-icon m-0" src={OpenInNewIcon} />
+          )}
+        </Stack>
+        <div>
+          <small className="clipboard-popover-detail-block-type">
+            {content.blockTypeDisplay}
+          </small>
+          <span>{intl.formatMessage(messages.popoverContentText)} </span>
```
Comment (ruzniaievdm) [inline src/course-unit/clipboard/paste-component/components/PopoverContent.jsx:33]: [nit] remove additional space

File: src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  const handleKeyDown = ({ key }) => {
+    if (key === 'Tab') {
+      popoverElementRef.current.focus();
+      handlePopoverToggle(true);
+    }
+  };
+
+  return (
+    <div
+      className="paste-component-whats-in-clipboard mt-2 d-flex align-items-center"
```
Comment (ruzniaievdm) [inline src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx:24]: [nit]
```suggestion
      className="paste-component-from-clipboard mt-2 d-flex align-items-center"
```

File: src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      data-testid="whats-in-clipboard"
```
Comment (ruzniaievdm) [inline src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx:25]: [nit] the same name as class above

Comment (ruzniaievdm) [reply src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx:25]: [nit] please check and below too

File: src/course-unit/data/thunk.js [hunk 6]
```diff
@@ -227,9 +239,8 @@ export function duplicateUnitItemQuery(itemId, xblockId) {
 
 export function copyToClipboard(usageKey) {
   const POLL_INTERVAL_MS = 1000; // Timeout duration for polling in milliseconds
-
+  // console.log('copyToClipboard ===========================');
```
Comment (ruzniaievdm) [inline src/course-unit/data/thunk.js:242]: please remove extra logs

File: src/generic/configure-modal/ConfigureModal.test.jsx [hunk 7]
```diff
@@ -65,7 +65,7 @@ describe('<ConfigureModal /> for Section', () => {
     axiosMock = new MockAdapter(getAuthenticatedHttpClient());
   });
 
-  it('renders ConfigureModal component correctly', () => {
+  it.skip('renders ConfigureModal component correctly', () => {
```
Comment (ruzniaievdm) [inline src/generic/configure-modal/ConfigureModal.test.jsx:68]: [question] Do we need to skip these tests?

File: src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx (hunk 4 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx:24]: I decided to remove the prefix "paste-component" and take the name of the component as a basis

File: src/course-unit/clipboard/paste-component/components/PopoverContent.jsx (hunk 3 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/clipboard/paste-component/components/PopoverContent.jsx:33]: Removed

File: src/course-unit/clipboard/paste-component/components/PasteComponentButton.jsx (hunk 2 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/clipboard/paste-component/components/PasteComponentButton.jsx:13]: Good catch!

File: src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx (hunk 5 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/clipboard/paste-component/components/WhatsInClipboard.jsx:25]: Thanks! I decided to remove the prefix "paste-component" and take the name of the component as a basis

File: src/course-unit/data/thunk.js (hunk 6 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/data/thunk.js:242]: Removed

File: src/generic/configure-modal/ConfigureModal.test.jsx (hunk 7 above)
Comment (PKulkoRaccoonGang) [reply src/generic/configure-modal/ConfigureModal.test.jsx:68]: Not anymore, @ihor-romaniuk promptly found and fixed the problem in https://github.com/raccoongang/frontend-app-course-authoring/pull/152

File: src/course-unit/__mocks__/clipboardResponse.js (hunk 1 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/__mocks__/clipboardResponse.js:4]: Nope, changed

**Review by ruzniaievdm — APPROVED** [review]

File: src/course-unit/CourseUnit.test.jsx [hunk 8]
```diff
@@ -668,4 +686,389 @@ describe('<CourseUnit />', () => {
...
+      const updatedCourseSectionVerticalData = cloneDeep(courseSectionVerticalMock);
+      const updatedAncestorsChild = updatedCourseSectionVerticalData.xblock_info.ancestor_info.ancestors[0];
+      set(updatedCourseSectionVerticalData, 'xblock_info.ancestor_info.ancestors[0].child_info.children', [
+        ...updatedAncestorsChild.child_info.children,
+        courseUnitMock,
+      ]);
+
+      units = getAllByTestId('course-unit-btn');
+      const courseUnits = courseSectionVerticalMock.xblock_info.ancestor_info.ancestors[0].child_info.children;
+      expect(units.length).toEqual(courseUnits.length);
```
Comment (ihor-romaniuk) [inline src/course-unit/CourseUnit.test.jsx:861]: [optional] this case and below can be replaced with
```suggestion
      expect(units).toHaveLength(courseUnits.length);
```

File: src/course-unit/clipboard/paste-component/constants.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import PropTypes from 'prop-types';
+
+export const clipboardProps = {
```
Comment (ihor-romaniuk) [inline src/course-unit/clipboard/paste-component/constants.js:3]: [question] What do you think about renaming it to `clipboardPropsTypes`?

File: src/course-unit/clipboard/paste-notification/index.jsx [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+import AlertMessage from '../../../generic/alert-message';
+import { ActionButton, AlertContent } from './components';
+import { getAlertStatus } from './utils';
+import { initialNotificationAlertsState } from './constants';
+import messages from './messages';
+
+const PastNotificationAlert = ({
+  staticFileNotices,
+  courseId,
+}) => {
```
Comment (ihor-romaniuk) [inline src/course-unit/clipboard/paste-notification/index.jsx:19]: [optional] Can be simplified to one line
```suggestion
const PastNotificationAlert = ({ staticFileNotices, courseId }) => {
```

File: src/course-unit/data/api.js [hunk 11]
```diff
@@ -64,17 +64,19 @@ export async function getCourseSectionVerticalData(unitId) {
  * @param {string} options.parentLocator - The parent locator of the XBlock.
  * @param {string} [options.displayName] - The display name for the XBlock.
  * @param {string} [options.boilerplate] - The boilerplate for the XBlock.
+ * @param {string} [options.stagedContent] - The staged content for the XBlock.
```
Comment (ihor-romaniuk) [inline src/course-unit/data/api.js:67]: [optional] Repeated parts may be omitted.
```suggestion
 * @param {string} options.parentLocator - The parent locator.
 * @param {string} [options.displayName] - The display name.
 * @param {string} [options.boilerplate] - The boilerplate.
 * @param {string} [options.stagedContent] - The staged content.
```

File: src/course-unit/hooks.jsx [hunk 12]
```diff
@@ -20,8 +20,11 @@ import {
   getLoadingStatus,
   getSavingStatus,
   getSequenceStatus,
+  getCourseUnitEnableCopyPaste, getStaticFileNotices,
```
Comment (ihor-romaniuk) [inline src/course-unit/hooks.jsx:23]: [nit] Place one property per line 
```suggestion
  getCourseUnitEnableCopyPaste,
  getStaticFileNotices,
```

File: src/course-unit/clipboard/paste-component/constants.js (hunk 9 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/clipboard/paste-component/constants.js:3]: Good idea

File: src/course-unit/data/api.js (hunk 11 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/data/api.js:67]: Done

File: src/course-unit/hooks.jsx (hunk 12 above)
Comment (PKulkoRaccoonGang) [reply src/course-unit/hooks.jsx:23]: Corrected, thanks

**Review by ihor-romaniuk — APPROVED** [review]

**Pull request merged**